### PR TITLE
feat(erb-parser): parse erb data to csv

### PR DESCRIPTION
**Pull request by neelk07**
What: 
 - a module to parse the latest erb data from a local text file downloaded from (https://www.frbservices.org/EPaymentsDirectory/FedACHdir.txt)

Why:
 - to validate routing numbers when creating bank accounts

Details:
- download the latest text file (https://www.frbservices.org/EPaymentsDirectory/FedACHdir.txt) and place anywhere in the root folder
- to test you can run 
```javascript 
node index.js -i <path/to/FedACHdir.txt> -o <path/to/save/output.csv> -c config/fedACHConfig.js
``` 
- create local table `banks` with 
```sql
create table banks (
  routing_number varchar(20),
  servicing_frb_number varchar(20),
  date_last_change date,
  customer_name varchar(40),
  address varchar(40),
  city varchar(40),
  state varchar(20),
  zipcode int,
  zipcode_extension int,
  telephone varchar(10),
  institution_status_code int,
  is_federal_reserve_bank boolean,
  office_type varchar(20)
);
```
 - import csv to db (psql lob_local)
```sql 
\copy banks from /path/to/erb-parser/output.csv DELIMITER ',' CSV HEADER;
```
- mess around with the data!

Testing:

- run `npm test` to test the package

fixes: https://github.com/lob/lob-api/issues/1896

@pon @mgartner @robinjoseph08 @elnaz @dmlittle

### CODE REVIEW

File: lib/to_csv.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+'use strict';
+
+var path     = require('path');
+var _isArray = require('lodash.isarray');
```
Comment (robinjoseph08) [inline lib/to_csv.js]: are you using this?

Comment (neelk07) [inline lib/to_csv.js]: @robinjoseph08 for some reason, when I was running it yesterday it told me that it required `lodash.isarray` I'll run it again without it and see if it works

File: lib/to_csv.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
+'use strict';
+
+var path     = require('path');
+// var _isArray = require('lodash.isarray');
```
Comment (elnaz) [inline lib/to_csv.js]: rogue line

**Comment by elnaz** [comment]
Great PR description! Could you also update the repo README with instructions for the code you've written so far in this PR?

File: lib/to_csv.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  bank.funds_transfer_eligible = bank.funds_transfer_eligible === 'Y' ? 't' : 'f';
+  bank.funds_settlement_only = bank.funds_settlement_only === 'S' ? 't' : 'f';
+  bank.book_entry_securities_transfer_eligible = bank.book_entry_securities_transfer_eligible === 'Y' ? 't' : 'f';
+  if (bank.date_last_revision) {
+    bank.date_last_revision = bank.date_last_revision.slice(0,4) + '-' + bank.date_last_revision.slice(4,6) + '-' + bank.date_last_revision.slice(6,8);
+  }
+  return bank;
+})
+.then(function (banks) {
+  return json2csv({ data: banks, fields: Object.keys(banks[0]) });
```
Comment (elnaz) [inline lib/to_csv.js]: Is it possible for banks[0] to not have a date_last_revision key? If so, then would that cause the resulting csv to not have a date_last_revision column?

**Comment by elnaz** [comment]
Done with comments. For anyone else that wants to run it locally, here's an erb_data folder you can drop in your project root before you run `node lib/to_csv.js erb_data/fpddir.txt output.csv`
[erb_data.zip | uploaded via ZenHub](https://files.zenhub.io/55aab1be466fe68f139b98e2)

File: package.json [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  "keywords": [
+    "banking"
+  ],
+  "author": "neel@lob.com",
+  "license": "ISC",
+  "dependencies": {
+    "bluebird": "^2.9.34",
+    "json2csv": "^2.8.0",
+    "request": "^2.58.0"
```
Comment (dmlittle) [inline package.json]: stranded dependency?

**Comment by mgartner** [comment]
I believe @neelk07 is going to try parsing another file because this list was incomplete. I'll mark this as WIP for now.

**Comment by neelk07** [comment]
@elnaz updated the README and also added the erb_data folder with default txt file for easier testing

**Comment by neelk07** [comment]
@pon @mgartner @robinjoseph08 @dmlittle 
ready for review again!

**Comment by dmlittle** [comment]
[LGTM](http://lgtm.in/i/mZVllIoaq)

**Comment by neelk07** [comment]
Changed back to WIP because of DB dump problem

Details:
 - the data has multiple rows for the same bank (with different last change dates) so when inserting in the database it violates a unique key constraint (routing_number) so I'm going to clean the data before righting out to file

**Comment by neelk07** [comment]
Fixed above problem by using only latest bank information. Ready for CR!

@pon @mgartner @elnaz @robinjoseph08 @leore

File: lib/to_csv.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
+'use strict';
+
+var path     = require('path');
+var Bluebird = require('bluebird');
+var fs       = Bluebird.promisifyAll(require('fs'));
+var json2csv = Bluebird.promisify(require('json2csv'));
+
+if (!process.argv[2] || process.argv[3] !== '--config' || ['wire','ach'].indexOf(process.argv[4]) === -1 ||
```
Comment (pon) [inline lib/to_csv.js]: You can clean a lot of this up with: https://www.npmjs.com/package/yargs

**Comment by pon** [comment]
I would split up the module into the following structure:

`lib/index.js` - required `to_csv` handle the arg parsing and execute the function
`lib/to_csv.js` - exports the `to_csv` function

**Comment by robinjoseph08** [comment]
do we care about testing and linting for this repo?

**Comment by mgartner** [comment]
Why do you need the `--config ach` option? What does that do? What other things can you put in other than ach?

**Comment by elnaz** [comment]
Is this WIP or ready for review?

**Comment by neelk07** [comment]
@mgartner we got rid of that to make it more extendable, new update should explain it all 

@elnaz should be ready for review later today!

**Comment by neelk07** [comment]
@pon @mgartner @robinjoseph08 @elnaz @dmlittle

Ready for review again!

File: test/to_csv.test.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    });
+
+    it('should create date of last change', function () {
+      expect(bankMap['999999999'].date_last_change).to.eql('2011-07-28');
+      expect(bankMap['011000015'].date_last_change).to.eql('2002-02-08');
+      expect(bankMap['011000138'].date_last_change).to.eql('2010-10-13');
+      expect(bankMap['011000536'].date_last_change).to.eql('2011-11-14');
+    });
+
+     it('should use newest date of last change', function () {
```
Comment (robinjoseph08) [inline test/to_csv.test.js]: indentation

File: test/to_csv.test.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+  describe('writeCSV', function () {
+
+    var testBankMap = { 0: { a: 'a', b: 'b' }, 1: { a: 'b', b: 'a' } };
+    var testBankCSV = '"a","b"\n"a","b"\n"b","a"';
+
+    it('should write to csv', function (done) {
+      var fsMock = sinon.mock(fs);
+      fsMock.expects('writeFileAsync').withExactArgs(outputPath, testBankCSV);
+      // fsMock.expects('readFileAsync').withExactArgs(outputPath).returns(testBankCSV)
```
Comment (robinjoseph08) [inline test/to_csv.test.js]: remove comment

**Comment by neelk07** [comment]
@robinjoseph08 made the fixes!

**Comment by mgartner** [comment]
LGTM

**Comment by robinjoseph08** [comment]
LGTM

**Pull request merged**